ccl/backupccl: TestBackupRestoreNegativePrimaryKey failed

[(ccl/backupccl).TestBackupRestoreNegativePrimaryKey failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1744976&tab=buildLog) on [master@afc1cb2923ec86cc274a0ef96b341d9fcbdd5b61](https://github.com/cockroachdb/cockroach/commits/afc1cb2923ec86cc274a0ef96b341d9fcbdd5b61):

Fatal error:
```
panic: interface conversion: jobspb.isPayload_Details is *jobspb.Payload_CreateStats, not *jobspb.Payload_Backup [recovered]
	panic: interface conversion: jobspb.isPayload_Details is *jobspb.Payload_CreateStats, not *jobspb.Payload_Backup [recovered]
	panic: interface conversion: jobspb.isPayload_Details is *jobspb.Payload_CreateStats, not *jobspb.Payload_Backup

```

Stack:
```
goroutine 38598 [running]:
testing.tRunner.func1(0xc003bb1f00)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x3922ac0, 0xc0049eaab0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:98 +0x210
panic(0x3922ac0, 0xc0049eaab0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/ccl/backupccl_test.backupAndRestore(0x45be8a0, 0xc000074120, 0xc003bb1f00, 0xc00105bab0, 0xc006236f90, 0x1, 0x1, 0xc006236fa0, 0x1, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup_test.go:515 +0x180c
github.com/cockroachdb/cockroach/pkg/ccl/backupccl_test.TestBackupRestoreNegativePrimaryKey(0xc003bb1f00)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup_test.go:435 +0x30d
testing.tRunner(0xc003bb1f00, 0x3f32ef8)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350

```

<details><summary>Log preceding fatal error</summary><p>

```
I200211 19:53:36.257316 49151 jobs/registry.go:656  [n1] job 528811347940933633: stepping through state running with error <nil>
I200211 19:53:36.319417 49151 jobs/registry.go:656  [n1] job 528811347940933633: stepping through state succeeded with error <nil>
E200211 19:53:36.331650 49151 jobs/registry.go:795  [n1] job 528811347940933633: status succeeded after adoption finished
I200211 19:53:36.408392 39667 gossip/gossip.go:567  [n2] gossip status (ok, 3 nodes)
gossip client (1/3 cur/max conns)
  1: 127.0.0.1:42147 (1m0s: infos 100/330 sent/received, bytes 89622B/184351B sent/received)
gossip server (0/3 cur/max conns, infos 100/330 sent/received, bytes 89622B/184351B sent/received)
gossip connectivity
  n1 [sentinel];
  n2 -> n1; n3 -> n1;
I200211 19:53:36.410522 39671 server/status/runtime.go:498  [n2] runtime stats: 285 MiB RSS, 765 goroutines, 75 MiB/98 MiB/123 MiB GO alloc/idle/total, 63 MiB/92 MiB CGO alloc/total, 615.7 CGO/sec, 7.8/1.2 %(u/s)time, 0.0 %gc (1x), 16 MiB/16 MiB (r/w)net
I200211 19:53:36.532789 39935 gossip/gossip.go:567  [n3] gossip status (ok, 3 nodes)
gossip client (1/3 cur/max conns)
  1: 127.0.0.1:42147 (1m0s: infos 120/312 sent/received, bytes 28673B/247078B sent/received)
gossip server (0/3 cur/max conns, infos 120/312 sent/received, bytes 28673B/247078B sent/received)
gossip connectivity
  n1 [sentinel];
  n2 -> n1; n3 -> n1;
I200211 19:53:36.537324 39939 server/status/runtime.go:498  [n3] runtime stats: 286 MiB RSS, 765 goroutines, 45 MiB/118 MiB/117 MiB GO alloc/idle/total, 63 MiB/92 MiB CGO alloc/total, 614.7 CGO/sec, 8.3/1.4 %(u/s)time, 0.0 %gc (2x), 16 MiB/16 MiB (r/w)net
I200211 19:53:38.943615 39673 sql/schema_changer.go:973  [n2,scExec] schema change on bank (53 v4) mutation 1 starting execution...
I200211 19:53:46.234099 38743 server/status/runtime.go:498  [n1] runtime stats: 288 MiB RSS, 757 goroutines, 60 MiB/106 MiB/114 MiB GO alloc/idle/total, 63 MiB/92 MiB CGO alloc/total, 410.6 CGO/sec, 8.1/1.4 %(u/s)time, 0.0 %gc (1x), 16 MiB/16 MiB (r/w)net
I200211 19:53:46.410743 39671 server/status/runtime.go:498  [n2] runtime stats: 289 MiB RSS, 757 goroutines, 68 MiB/99 MiB/114 MiB GO alloc/idle/total, 68 MiB/97 MiB CGO alloc/total, 250.6 CGO/sec, 5.7/1.2 %(u/s)time, 0.0 %gc (1x), 15 MiB/15 MiB (r/w)net
I200211 19:53:46.534540 39939 server/status/runtime.go:498  [n3] runtime stats: 290 MiB RSS, 757 goroutines, 40 MiB/123 MiB/119 MiB GO alloc/idle/total, 73 MiB/102 MiB CGO alloc/total, 251.4 CGO/sec, 5.8/1.1 %(u/s)time, 0.0 %gc (1x), 15 MiB/15 MiB (r/w)net
I200211 19:53:55.558712 45536 jobs/registry.go:656  [n1] job 528811200803340289: stepping through state succeeded with error <nil>
E200211 19:53:55.594727 45536 jobs/registry.go:795  [n1] job 528811200803340289: status succeeded after adoption finished
I200211 19:53:55.606439 41332 sql/event_log.go:132  [n1,client=127.0.0.1:60416,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.stats.automatic_collection.enabled Value:false User:root}
I200211 19:53:55.661784 38598 util/stop/stopper.go:539  quiescing
I200211 19:53:55.661839 50892 util/stop/stopper.go:539  quiescing
I200211 19:53:55.661853 50890 util/stop/stopper.go:539  quiescing
I200211 19:53:55.661955 50891 util/stop/stopper.go:539  quiescing
W200211 19:53:55.668288 39789 gossip/gossip.go:1518  [n3] no incoming or outgoing connections
W200211 19:53:55.668329 39452 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
W200211 19:53:55.668448 41646 storage/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 3: EOF:
W200211 19:53:55.668469 40180 storage/raft_transport.go:637  [n1] while processing outgoing Raft queue to node 3: EOF:
W200211 19:53:55.668477 42825 storage/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 3: EOF:
W200211 19:53:55.668513 40070 storage/raft_transport.go:637  [n1] while processing outgoing Raft queue to node 3: EOF:
W200211 19:53:55.668563 40092 storage/raft_transport.go:637  [n3] while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200211 19:53:55.668590 39990 storage/raft_transport.go:637  [n1] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200211 19:53:55.668727 40085 storage/raft_transport.go:637  [n3] while processing outgoing Raft queue to node 1: EOF:
W200211 19:53:55.668742 40179 storage/raft_transport.go:637  [n1] while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
W200211 19:53:55.668799 39997 storage/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 1: EOF:
W200211 19:53:55.668826 40005 storage/raft_transport.go:637  [n2] while processing outgoing Raft queue to node 1: EOF:
W200211 19:53:55.668942 42879 storage/raft_transport.go:637  [n3] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200211 19:53:55.668962 41872 storage/raft_transport.go:637  [n3] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200211 19:53:55.669055 51355 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I200211 19:53:55.732624 51370 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n2: failed to connect to n2 at 127.0.0.1:40809: context canceled
I200211 19:53:55.732654 51369 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n1: failed to connect to n1 at 127.0.0.1:42147: context canceled
I200211 19:53:55.808398 51268 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n1: failed to connect to n1 at 127.0.0.1:42147: context canceled
I200211 19:53:55.808398 51381 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n3: failed to connect to n3 at 127.0.0.1:44949: context canceled
--- FAIL: TestBackupRestoreNegativePrimaryKey (79.79s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreNegativePrimaryKey PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreNegativePrimaryKey.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -553,23 +553,36 @@ func backupAndRestore(
 			t.Fatalf("expected %d rows for %d accounts, got %d", expected, numAccounts, exported.rows)
 		}
 
-		sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=false`)
-		const stmt = "SELECT payload FROM system.jobs ORDER BY created DESC LIMIT 1"
-		var payloadBytes []byte
-		sqlDB.QueryRow(t, stmt).Scan(&payloadBytes)
+		found := false
+		const stmt = "SELECT payload FROM system.jobs ORDER BY created DESC LIMIT 10"
+		for rows := sqlDB.Query(t, stmt); rows.Next(); {
+			var payloadBytes []byte
+			if err := rows.Scan(&payloadBytes); err != nil {
+				t.Fatal(err)
+			}
 
-		payload := &jobspb.Payload{}
-		if err := protoutil.Unmarshal(payloadBytes, payload); err != nil {
-			t.Fatal("cannot unmarshal job payload from system.jobs")
-		}
+			payload := &jobspb.Payload{}
+			if err := protoutil.Unmarshal(payloadBytes, payload); err != nil {
+				t.Fatal("cannot unmarshal job payload from system.jobs")
+			}
 
-		backupManifest := &backupccl.BackupManifest{}
-		backupDetails := payload.Details.(*jobspb.Payload_Backup).Backup
-		if err := protoutil.Unmarshal(backupDetails.BackupManifest, backupManifest); err != nil {
-			t.Fatal("cannot unmarshal backup descriptor from job payload from system.jobs")
+			backupManifest := &backupccl.BackupManifest{}
+			backupPayload, ok := payload.Details.(*jobspb.Payload_Backup)
+			if !ok {
+				t.Logf("job %T is not a backup: %v", payload.Details, payload.Details)
+				continue
+			}
+			backupDetails := backupPayload.Backup
+			found = true
+			if err := protoutil.Unmarshal(backupDetails.BackupManifest, backupManifest); err != nil {
+				t.Fatal("cannot unmarshal backup descriptor from job payload from system.jobs")
+			}
+			if backupManifest.Statistics != nil {
+				t.Fatal("expected statistics field of backup descriptor payload to be nil")
+			}
 		}
-		if backupManifest.Statistics != nil {
-			t.Fatal("expected statistics field of backup descriptor payload to be nil")
+		if !found {
+			t.Fatal("scanned job rows did not contain a backup!")
 		}
 	}
 

